discover.py

```diff
@@ -0,0 +1,234 @@
+import logging
+import time
+import shutil
+import subprocess
+from threading import Thread
+from datetime import datetime
+
+from zeroconf import ServiceBrowser, Zeroconf, ServiceInfo
+import netifaces
+from flask import Flask
+from flask_jsontools import JsonSerializableBase, jsonapi
+from waitress import serve
+
+import supervisor_client
+
+log = logging.getLogger('discovery')
+
+SERVICE_TYPE = "_liquid._tcp.local."
+HOST_ONLY_NETWORK_MASK = '255.255.255.255'
+DNSMASQ_PROCESS_NAME = 'dnsmasq-dns'
+
+nodes = {}
+
+def reload_dnsmasq():
+    log.debug("Reloading dnsmasq-dns")
+    supervisor = supervisor_client.connect_unix_socket()
+    is_running = supervisor.getProcessInfo(DNSMASQ_PROCESS_NAME)['statename'] == 'RUNNING'
+    if is_running:
+        supervisor.stopProcess(DNSMASQ_PROCESS_NAME)
+    supervisor.startProcess(DNSMASQ_PROCESS_NAME)
+
+def stop_dnsmasq():
+    log.debug("Stopping dnsmasq-dns")
+    supervisor = supervisor_client.connect_unix_socket()
+    supervisor.stopProcess(DNSMSAQ_PROCESS_NAME)
+
+def rewrite_dnsmasq_conf(path, conf_string):
+    temp = path + ".tmp"
+    with open(temp, 'w') as f:
+        f.write(conf_string)
+    shutil.move(temp, path)
+
+class DnsmasqRestarter(Thread):
+    def __init__(self):
+        super().__init__()
+        self.last_dnsmasq_command = ''
+
+    def run(self):
+        while True:
+            self.update_dnsmasq()
+            time.sleep(3)
+
+    def update_dnsmasq(self):
+        interface = app.config['DNS_INTERFACE']
+        if not interface:
+            if self.last_dnsmasq_command:
+                stop_dnsmasq()
+                self.last_dnsmasq_command = ''
+            return
+
+        command = get_dns_conf_string(interface)
+        if command == self.last_dnsmasq_command:
+            return
+
+        rewrite_dnsmasq_conf(app.config['DNS_CONF_FILE'], command)
+        reload_dnsmasq()
+        self.last_dnsmasq_command = command
+
+
+def get_dns_conf_string(interface):
+    ip = get_ipv4_addr(interface)
+    common_args = [
+        'domain-needed',
+        'bogus-priv',
+        'server=208.67.222.222',
```

Comment:
We shouldn't be hardcoding DNS servers. At least make them configurable in settings.